samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -176,30 +181,38 @@ public void start() {
    *
    */
   public synchronized void stop() {
-    boolean containerShutdownInvoked = false;
     if (container != null) {
       try {
-        LOGGER.info("Shutting down container " + container.toString() + " from StreamProcessor");
+        LOGGER.info("Shutting down the container: {} of stream processor: {}.", container, processorId);
         container.shutdown();
-        containerShutdownInvoked = true;
-      } catch (IllegalContainerStateException icse) {
-        LOGGER.info("Container was not running", icse);
+        LOGGER.info("Waiting {} milliseconds for the container: {} to shutdown.", taskShutdownMs, container);
+        jcContainerShutdownLatch.await(taskShutdownMs, TimeUnit.MILLISECONDS);
+        if (executorService != null) {
+          executorService.shutdown();
+          executorService.awaitTermination(taskShutdownMs, TimeUnit.MILLISECONDS);
+          if (!executorService.isShutdown()) {
+            LOGGER.info("Forcefully shutting down the executor service of the container: {}.", container);
+            executorService.shutdownNow();
+          }
+        }
+      } catch (Exception exception) {
+        LOGGER.info(String.format("Shutdown of container: %s threw an exception.", container), exception);
       }
     }
+    stopJobCoordinator();
```

Comment:
this logic of shutting down container and jobCoordinator in different order becoming very complicated. Is there a way to simplify it?